### PR TITLE
Adding services.

### PR DESCRIPTION
**Pull request by Kotsiuba**

### CODE REVIEW

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+using System.Linq;
+using System.Text;
+using ToDo.DataAccess.Models;
+using ToDo.DataAccess.Repositories;
+using ToDo.Services.ViewModels;
+using ToDo.WebAPI.ViewModels;
+
+namespace ToDo.Services.Handlers
+{
+    public interface IItemHandler
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: interface into separate folder

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        IEnumerable<ItemDisplayViewModel> Search(string category, string[] tags);
+        void Create(ItemCreateViewModel value);
+        void Delete(Guid id);
+    }
+
+    public class ItemHandler : IItemHandler
+    {
+        private readonly IItemRepository _itemRepository;
+
+        public ItemHandler(IItemRepository itemRepository)
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: public var
private var
public prop
private prop
constructors
public methods
private methods

**Review by zhikin2207 — CHANGES_REQUESTED** [review]

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs (hunk 1 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Removed

File: ToDo.DataAccess/ToDo.DataAccess/Repositories/CustomRepositories/ItemRepository.cs [hunk 3]
```diff
@@ -1,23 +1,20 @@
...
-    public interface IItemRepository : IGenericRepository<Item>
-    { 
-        
-    }
-
     public class ItemRepository : GenericRepository<Item>, IItemRepository
     {
         public ItemRepository(ToDoDbContext context) : base(context)
         {
+            ToDoDbInitializer.Initialize(context);
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.DataAccess/Repositories/CustomRepositories/ItemRepository.cs]: WTF?

File: ToDo.DataAccess/ToDo.Services/DTOs/ItemDTO.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    public class ItemDTO
+    {
+        public Guid Id { get; set; }
+        public string Title { get; set; }
+        public string Description { get; set; }
+        public DateTime DueDate { get; set; }
+        public bool Status { get; set; }
+        public Priority Priority { get; set; }
+
+        public Category Category { get; set; }
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/DTOs/ItemDTO.cs]: Should contain only DTO objects.

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        public void Create(ItemDTO value)
+        {
+            _itemRepository.Add(ConvertToItem(value));
+        }
+
+        public void Delete(Guid id)
+        {
+            _itemRepository
+                .Delete(
+                    _itemRepository
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: into separate operators.

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        }
+
+        public ItemHandler(IItemRepository itemRepository)
+        {
+            _itemRepository = itemRepository;
+        }
+
+        public void Create(ItemDTO value)
+        {
+            _itemRepository.Add(ConvertToItem(value));
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Into separate operators.

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs (hunk 2 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Sorted.

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+                   .Intersect(tags)
+                   .Any())
+               .Select(ConvertToItemDisplayViewModel);
+        }
+
+        public StatisticDTO GetAdultItems()
+        {
+            Dictionary<string, int> counts = new Dictionary<string, int>();
+
+            foreach(var p in Enum.GetValues(typeof(Priority))
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Looks ugly, move enum items into the variable.

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+               .Where(i => i.TagItem
+                   .Select(ti => ti.Tag.Name)
+                   .Intersect(tags)
+                   .Any())
+               .Select(ConvertToItemDisplayViewModel);
+        }
+
+        public StatisticDTO GetAdultItems()
+        {
+            Dictionary<string, int> counts = new Dictionary<string, int>();
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Fuk name

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+                                        .Cast<Priority>()
+                                        .Select(v => v.ToString()))
+                                        
+            {
+                counts.Add(p, 0);
+            }
+
+            var suitableItems = _itemRepository
+                    .GetAll()
+                    .Where(i => i.Title.Contains("xxx")
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: looks ugly

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+                            || string.Equals(
+                               i.Category.Name,
+                               "adult",
+                               StringComparison.CurrentCultureIgnoreCase)
+                            || i.Description.Contains("xxx"))
+                    .Where(i => i.Status == false)
+                    .OrderBy(i => i.DueDate)
+                    .Select(ConvertToItemDisplayViewModel);
+
+            foreach(var i in suitableItems)
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: looks ugly

File: ToDo.DataAccess/ToDo.WebAPI/ViewModels/ItemCreateViewModel.cs [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+    public class ItemCreateViewModel
+    {
+        public Guid Id { get; set; }
+        public string Title { get; set; }
+        public string Description { get; set; }
+        public DateTime DueDate { get; set; }
+        public bool Status { get; set; }
+        public Priority Priority { get; set; }
+
+        public Category Category { get; set; }
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.WebAPI/ViewModels/ItemCreateViewModel.cs]: WTF?

File: ToDo.DataAccess/ToDo.WebAPI/ViewModels/StatisticViewModel.cs [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+using System.Threading.Tasks;
+using ToDo.Services.DTOs;
+
+namespace ToDo.WebAPI.ViewModels
+{
+    public class StatisticViewModel
+    {
+        public IEnumerable<ItemDisplayViewModel> items { get; set; }
+
+        public Dictionary<string, int> priorityCounts { get; set; }
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.WebAPI/ViewModels/StatisticViewModel.cs]: Why names are lowercase?

**Review by zhikin2207 — CHANGES_REQUESTED** [review]

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs (hunk 8 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: renamed to newPriorityCounts

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs (hunk 7 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Moved.

File: ToDo.DataAccess/ToDo.DataAccess/Repositories/CustomRepositories/ItemRepository.cs (hunk 3 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.DataAccess/Repositories/CustomRepositories/ItemRepository.cs]: I recreated base and forgot  to delete it.

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs (hunk 5 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Separeted.

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs (hunk 6 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Separeted.

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs (hunk 9 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Changed.

File: ToDo.DataAccess/ToDo.WebAPI/ViewModels/StatisticViewModel.cs (hunk 12 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.WebAPI/ViewModels/StatisticViewModel.cs]: Changed.

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs (hunk 10 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Changed.

File: ToDo.DataAccess/ToDo.Services/DTOs/ItemDTO.cs [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+namespace ToDo.Services.DTOs
+{
+    public class ItemDTO
+    {
+        public Guid Id { get; set; }
+        public string Title { get; set; }
+        public string Description { get; set; }
+        public DateTime DueDate { get; set; }
+        public bool Status { get; set; }
+        public Priority Priority { get; set; }
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/DTOs/ItemDTO.cs]: Priority?

File: ToDo.DataAccess/ToDo.Services/DTOs/TagItemDTO.cs [hunk 14]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+
+namespace ToDo.Services.DTOs
+{
+    public class TagItemDTO
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/DTOs/TagItemDTO.cs]: You don't need to show this table out of services.

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        }
+
+        public StatisticDTO GetAdultItems()
+        {
+            Dictionary<string, int> newPriorityCounts = new Dictionary<string, int>();
+            var listOfPriorities = Enum.GetValues(typeof(Priority))
+                                        .Cast<Priority>()
+                                        .Select(v => v.ToString());
+
+            foreach (var p in listOfPriorities)
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: p?

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+                                        
+            {
+                newPriorityCounts.Add(p, 0);
+            }
+
+            Func<Item, bool> condition = (i => (i.Title.Contains("xxx")
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: No, still ugly. And don't move this into separate function

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+
+            var suitableItems = _itemRepository
+                    .GetAll()
+                    .Where(condition)
+                    .OrderBy(i => i.DueDate)
+                    .Select(HandlerConverter.ConvertToItemDTO);
+
+            foreach(var i in suitableItems)
+            {
+                var key = i.Priority.ToString();
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: i?

File: ToDo.DataAccess/ToDo.Tests/Testing.cs [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+using NUnit.Framework;
+using System;
+using ToDo.DataAccess.Models;
+using ToDo.DataAccess.Repositories.CustomRepositories;
+using ToDo.Services.Handlers;
+
+namespace ToDo.Tests
+{
+    [TestFixture]
+    public class Testing
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Tests/Testing.cs:15]: No tests in this PR

File: ToDo.DataAccess/ToDo.Services/DTOs/ItemDTO.cs (hunk 4 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/DTOs/ItemDTO.cs]: Changed

**Review by zhikin2207 — CHANGES_REQUESTED** [review]

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs (hunk 5 above)
Comment (zhikin2207) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: not separate lines, but separate operators

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs (hunk 6 above)
Comment (zhikin2207) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: still not

File: ToDo.DataAccess/ToDo.WebAPI/ViewModels/ItemCreateViewModel.cs (hunk 11 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.WebAPI/ViewModels/ItemCreateViewModel.cs]: Changed.

File: ToDo.DataAccess/ToDo.Services/DTOs/ItemDTO.cs (hunk 13 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/DTOs/ItemDTO.cs]: Changed to PriorotyDTO

File: ToDo.DataAccess/ToDo.Services/DTOs/TagItemDTO.cs (hunk 14 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/DTOs/TagItemDTO.cs]: Deleted.

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs (hunk 15 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Changed.

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs (hunk 16 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Moved into function.

Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: As a variant*

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs (hunk 17 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: changed.

File: ToDo.DataAccess/ToDo.Tests/Testing.cs (hunk 18 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Tests/Testing.cs:15]: Deleted

File: ToDo.DataAccess/ToDo.Services/Configuration/MappingServicesProfile.cs [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+using ToDo.Services.DTOs;
+
+namespace ToDo.Services.Configuration
+{
+    public class MappingServicesProfile : Profile
+    {
+        public MappingServicesProfile()
+        {
+            CreateMap<CategoryDTO, Category>()
+                               .ForMember<Category>(c => c.Parent, opt => opt.MapFrom(c =>
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Configuration/MappingServicesProfile.cs]: Fix tabs

File: ToDo.DataAccess/ToDo.Services/Configuration/MappingServicesProfile.cs [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+                        (c.Parent == null) ? null
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Configuration/MappingServicesProfile.cs]: Remove these lines.

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+            {
+                tagItem.Add(
+                    new TagItem
+                    {
+                        Item = item,
+                        ItemId = item.Id,
+                        Tag = tag,
+                        TagId = tag.Id
+                    }
+                    );
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Fix spacing

```
tagItem.Add(new TagItem
{
    Item = item,
    ItemId = item.Id,
    Tag = tag,
    TagId = tag.Id
});
```

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+            }
+
+            item.TagItem = tagItem;
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: `item.TagItem = tags.Select(t => new TagItem ...`

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+                   .Intersect(tags)
+                   .Any())
+               .Select(_mapper.Map<Item, ItemDTO>);
+        }
+
+        public StatisticDTO GetAdultItems()
+        {
+            Dictionary<string, int> newPriorityCounts = new Dictionary<string, int>();
+            var listOfPriorities = Enum.GetValues(typeof(Priority))
+                                        .Cast<Priority>()
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Fix Spacing

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+            return new StatisticDTO
+            {
+                Items = suitableItems,
+                PriorityCounts = newPriorityCounts
+            };
+        }
+
+        public bool IsItemAdult(Item i)
+        {
+            return (i.Title.Contains("xxx")
```
Comment (zhikin2207) [inline ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs:105]: Will fix with me.

File: ToDo.DataAccess/ToDo.Tests/Testing.cs (hunk 18 above)
Comment (zhikin2207) [reply ToDo.DataAccess/ToDo.Tests/Testing.cs:15]: WTF? What is deleted?

**Review by zhikin2207 — CHANGES_REQUESTED** [review]

File: ToDo.DataAccess/ToDo.Services/Configuration/MappingServicesProfile.cs (hunk 19 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/Configuration/MappingServicesProfile.cs]: Fixed.

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs (hunk 22 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Looks better

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs (hunk 21 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Fixed

File: ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs (hunk 23 above)
Comment (Kotsiuba) [reply ToDo.DataAccess/ToDo.Services/Handlers/ItemHandler.cs]: Fixed.

**Pull request merged**